File: src/main/java/com/sudosoo/takeiteasy/service/HeartServiceImpl.java
package com.sudosoo.takeiteasy.service;

import com.sudosoo.takeiteasy.common.service.JpaService;
import com.sudosoo.takeiteasy.dto.heart.CommentHeartRequestDto;
import com.sudosoo.takeiteasy.dto.heart.PostHeartRequestDto;
import com.sudosoo.takeiteasy.entity.Comment;
import com.sudosoo.takeiteasy.entity.Heart;
import com.sudosoo.takeiteasy.entity.Post;
import com.sudosoo.takeiteasy.repository.HeartRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Transactional
public class HeartServiceImpl implements HeartService, JpaService<Heart,Long> {

    private final HeartRepository heartRepository;
    private final PostService postService;
    private final CommentService commentService;

    @Override
    public JpaRepository<Heart, Long> getJpaRepository() {
        return heartRepository;
    }

    @Override
    public Heart createPostHeart(PostHeartRequestDto requestDto){
        Long memberId = requestDto.getMemberId();
        Post post = postService.getByPostId(requestDto.getPostId());
        if( heartRepository.existsByMemberIdAndPost(memberId, post)){
            throw new IllegalArgumentException("Duplicated Like !");
        }

        Heart heart = Heart.getPostHeart(post,memberId);

        return save(heart);
    }

    @Override
    public Heart createCommentHeart(CommentHeartRequestDto requestDto)  {
        Long memberId = requestDto.getMemberId();
        Comment comment  = commentService.getByCommentId(requestDto.getCommentId());
        if(heartRepository.existsByMemberIdAndComment(memberId, comment)){
            throw new IllegalArgumentException("Duplicated Like !");
        }

        Heart heart = Heart.getCommentHeart(comment,memberId);

        return save(heart);
    }


    @Override
    public void postDisHeart(PostHeartRequestDto heartRequestDTO) {
        Long memberId = heartRequestDTO.getMemberId();
        Post post = postService.getByPostId(heartRequestDTO.getPostId());
        Heart heart = findHeartByMemberAndPostOrComment(memberId,post);

        heart.unHeartPost();

        deleteById(heart.getId());
    }

    @Override
    public void commentDisHeart(CommentHeartRequestDto heartRequestDTO) {
        Long memberId =heartRequestDTO.getMemberId();
        Comment comment  = commentService.getByCommentId(heartRequestDTO.getCommentId());
        Heart heart = findHeartByMemberAndPostOrComment(memberId, comment);

        heart.unHeartComment();
        //TODO deleteById null 체크 뺄지 확인
        deleteById(heart.getId());
    }



    private Heart findHeartByMemberAndPostOrComment(Long memberId, Object reference) {
        if (reference instanceof Post) {
            return getPostHeart(memberId,(Post)reference);
        } else if (reference instanceof Comment) {
            return getCommentHeart(memberId,(Comment)reference);
        } else {
            throw new IllegalArgumentException("Unsupported reference type");
        }
    }

    private Heart getPostHeart(Long memberId ,Post p){
        return heartRepository.findByMemberIdAndPost(memberId,p)
                .orElseThrow(() -> new IllegalArgumentException("Could not find PostHeart id"));
    }

    private Heart getCommentHeart(Long memberId ,Comment c){
        return heartRepository.findByMemberIdAndComment(memberId,c)
                .orElseThrow(() -> new IllegalArgumentException("Could not find CommentHeart id"));
    }

}

File: takeItEasy-application/src/main/kotlin/com/sudosoo/takeItEasy/service/HeartServiceImpl.kt
package com.sudosoo.takeItEasy.service

import com.sudosoo.takeItEasy.dto.heart.CommentHeartRequestDto
import com.sudosoo.takeItEasy.dto.heart.PostHeartRequestDto
import com.sudosoo.takeiteasy.entity.Comment
import com.sudosoo.takeiteasy.entity.Heart
import com.sudosoo.takeiteasy.entity.Post
import com.sudosoo.takeiteasy.repository.HeartRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class HeartServiceImpl(
    val heartRepository: HeartRepository,
    val postService: PostService,
    val commentService: CommentService ) : HeartService {
    override fun createPostHeart(requestDto: PostHeartRequestDto): Heart {
        //TODO MemberSetting
        val memberId: Long = requestDto.memberId
        val post: Post = postService.getByPostId(requestDto.postId)

        require(!heartRepository.existsByMemberIdAndPost(memberId, post)) { "Duplicated Like !" }
        val heart: Heart = Heart.getPostHeart(post, memberId)

        return heartRepository.save<Heart>(heart)
    }

    override fun createCommentHeart(requestDto: CommentHeartRequestDto): Heart {
        //TODO MemberSetting
        val memberId: Long = requestDto.memberId
        val comment = commentService.getByCommentId(requestDto.commentId)

        require(!heartRepository.existsByMemberIdAndComment(memberId, comment)) { "Duplicated Like !" }
        val heart: Heart = Heart.getCommentHeart(comment, memberId)

        return heartRepository.save<Heart>(heart)
    }


    override fun postDisHeart(heartRequestDTO: PostHeartRequestDto) {
        //TODO MemberSetting
        val memberId: Long = heartRequestDTO.memberId
        val post: Post = postService.getByPostId(heartRequestDTO.postId)
        val heart: Heart = findHeartByMemberAndPostOrComment(memberId, post)

        heart.unHeartPost()

        heartRepository.delete(heart)
    }

    override fun commentDisHeart(heartRequestDTO: CommentHeartRequestDto) {
        //TODO MemberSetting
        val memberId: Long = heartRequestDTO.memberId
        val comment = commentService.getByCommentId(heartRequestDTO.commentId)
        val heart: Heart = findHeartByMemberAndPostOrComment(memberId, comment)

        heart.unHeartComment()

        heartRepository.delete(heart)
    }

    private fun findHeartByMemberAndPostOrComment(memberId: Long, reference: Any): Heart {
        return when (reference) {
            is Post -> {
                getPostHeart(memberId, reference)
            }
            is Comment -> {
                getCommentHeart(memberId, reference)
            }
            else -> {
                throw IllegalArgumentException("Unsupported reference type")
            }
        }
    }

    private fun getPostHeart(memberId: Long, p: Post?): Heart {
        return heartRepository.findByMemberIdAndPost(memberId, p)
            .orElseThrow{ IllegalArgumentException("Could not find PostHeart id") }
    }

    private fun getCommentHeart(memberId: Long, c: Comment): Heart {
        return heartRepository.findByMemberIdAndComment(memberId, c)
            .orElseThrow{ IllegalArgumentException("Could not find CommentHeart id") }
    }
}
